Amesos2 KLU2 does not build with -DTeuchos_ENABLE_COMPLEX:BOOL=ON

@trilinos/amesos2  The Amesos2 KLU2 solver fails to build with complex arithmetic enabled.  A very simple build of Amesos2 with KLU2 enabled, using this configure script:

cmake \
-DCMAKE_C_COMPILER=gcc-4 \
-DCMAKE_CXX_COMPILER=g++-4 \
-DCMAKE_Fortran_COMPILER=gfortran \
-D Trilinos_ENABLE_Amesos2:BOOL=ON \
  -D Amesos2_ENABLE_KLU2:BOOL=ON \
-D Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=OFF \
-D Trilinos_ENABLE_Tpetra:BOOL=ON \
-D TPL_ENABLE_LAPACK:BOOL=ON \
-D TPL_ENABLE_BLAS:BOOL=ON \
-D Trilinos_ENABLE_COMPLEX:BOOL=ON \
-D Teuchos_ENABLE_COMPLEX:BOOL=ON \
-D CMAKE_VERBOSE_MAKEFILE:BOOL=ON \
-D CMAKE_BUILD_TYPE:STRING=DEBUG \
-D CMAKE_CXX_FLAGS="-Wall -D_GLIBCXX_DEBUG" \
-D Trilinos_ENABLE_EXAMPLES:BOOL=ON \
-D Trilinos_ENABLE_TESTS:BOOL=ON \
../

Will result in a build error:

/Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_KLU2_def.hpp: In instantiation of ‘int Amesos2::KLU2< <template-parameter-1-1>, <template-parameter-1-2> >::numericFactorization_impl() [with Matrix = Tpetra::CrsMatrix<std::complex<double>, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial>, false>; Vector = Tpetra::MultiVector<std::complex<double>, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >]’:
/Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_SolverCore_def.hpp:153:3:   required from ‘Amesos2::Solver<Matrix, Vector>& Amesos2::SolverCore<ConcreteSolver, Matrix, Vector>::numericFactorization() [with ConcreteSolver = Amesos2::KLU2; Matrix = Tpetra::CrsMatrix<std::complex<double>, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial>, false>; Vector = Tpetra::MultiVector<std::complex<double>, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >]’
/Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_Details_registerLinearSolverFactory.cpp:102:1:   required from here
/Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_KLU2_def.hpp:247:45: error: no matching function for call to ‘klu_factor(int*, int*&, complex_type*&, KLU2::klu_symbolic<std::complex<double>, int>*&, KLU2::klu_common<std::complex<double>, int>*)’
            data_.symbolic_, &(data_.common_)) ;
                                             ^
/Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_KLU2_def.hpp:247:45: note: candidate is:
In file included from /Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/KLU2/Include/klu2_internal.h:48:0,
                 from /Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/KLU2/Source/klu2_defaults.hpp:41,
                 from /Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_KLU2_FunctionMap.hpp:68,
                 from /Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_KLU2_decl.hpp:57,
                 from /Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_KLU2.hpp:47,
                 from /Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_Factory.hpp:99,
                 from /Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_Details_LinearSolverFactory_def.hpp:52,
                 from /Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_Details_LinearSolverFactory.hpp:49,
                 from /Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_Details_registerLinearSolverFactory.cpp:45:
/Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/KLU2/Include/klu2_version.h:79:20: note: template<class Entry, class Int> KLU2::klu_numeric<Entry, Int>* KLU2::klu_factor(Int*, Int*, Entry*, KLU2::klu_symbolic<Entry, Int>*, KLU2::klu_common<Entry, Int>*)
 #define KLU_factor klu_factor
                    ^
/Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/KLU2/Source/klu2_factor.hpp:428:26: note: in expansion of macro ‘KLU_factor’
 KLU_numeric<Entry, Int> *KLU_factor         /* returns NULL if error, or a valid
                          ^
/Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/KLU2/Include/klu2_version.h:79:20: note:   template argument deduction/substitution failed:
 #define KLU_factor klu_factor
                    ^
/Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/KLU2/Source/klu2_factor.hpp:428:26: note: in expansion of macro ‘KLU_factor’
 KLU_numeric<Entry, Int> *KLU_factor         /* returns NULL if error, or a valid
                          ^
In file included from /Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_KLU2.hpp:49:0,
                 from /Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_Factory.hpp:99,
                 from /Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_Details_LinearSolverFactory_def.hpp:52,
                 from /Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_Details_LinearSolverFactory.hpp:49,
                 from /Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_Details_registerLinearSolverFactory.cpp:45:
/Users/hkthorn/Trilinos/Trilinos/packages/amesos2/src/Amesos2_KLU2_def.hpp:247:45: note:   cannot convert ‘sp_values’ (type ‘complex_type* {aka Kokkos::complex<double>*}’) to type ‘std::complex<double>*’
            data_.symbolic_, &(data_.common_)) ;
                                             ^

--- a/packages/amesos2/cmake/Dependencies.cmake
+++ b/packages/amesos2/cmake/Dependencies.cmake
@@ -2,8 +2,8 @@
 # Need a flag for LGPL vs BSD, SuperLU is required only for BSD.
 #
 # SET(LIB_REQUIRED_DEP_PACKAGES Teuchos Tpetra KLU2)
-SET(LIB_REQUIRED_DEP_PACKAGES Teuchos Tpetra TrilinosSS)
-SET(LIB_OPTIONAL_DEP_PACKAGES Epetra EpetraExt ShyLU_NodeBasker ShyLU_NodeTacho Kokkos)
+SET(LIB_REQUIRED_DEP_PACKAGES Teuchos Tpetra TrilinosSS Kokkos)
+SET(LIB_OPTIONAL_DEP_PACKAGES Epetra EpetraExt ShyLU_NodeBasker ShyLU_NodeTacho)
 SET(TEST_REQUIRED_DEP_PACKAGES)
 SET(TEST_OPTIONAL_DEP_PACKAGES ShyLU_NodeBasker ShyLU_NodeTacho Kokkos TrilinosSS)
 # SET(LIB_REQUIRED_DEP_TPLS SuperLU)

--- a/packages/amesos2/src/Amesos2_KLU2.cpp
+++ b/packages/amesos2/src/Amesos2_KLU2.cpp
@@ -45,8 +45,8 @@
 
 #ifdef HAVE_AMESOS2_EXPLICIT_INSTANTIATION
 
-#  include "Amesos2_KLU2_def.hpp"
-#  include "Amesos2_ExplicitInstantiationHelpers.hpp"
+#include "Amesos2_KLU2_def.hpp"
+#include "Amesos2_ExplicitInstantiationHelpers.hpp"
 
 namespace Amesos2 {
 

--- a/packages/amesos2/src/Amesos2_Util.hpp
+++ b/packages/amesos2/src/Amesos2_Util.hpp
@@ -220,13 +220,15 @@ namespace Amesos2 {
     template < class T0, class T1 >
     struct getStdCplxType
     {
-      using type = T0;
+      using common_type = typename std::common_type<T0,T1>::type;
+      using type = common_type;
     };
 
     template < class T0, class T1 >
     struct getStdCplxType< T0, T1* >
     {
-      using type = T1;
+      using common_type = typename std::common_type<T0,T1>::type;
+      using type = common_type;
     };
 
 #if defined(HAVE_TEUCHOS_COMPLEX) && defined(HAVE_AMESOS2_KOKKOS)
@@ -234,6 +236,13 @@ namespace Amesos2 {
     struct getStdCplxType< T0, Kokkos::complex<T0>* >
     {
       using type = std::complex<T0>;
+    };
+
+    template < class T0 , class T1 >
+    struct getStdCplxType< T0, Kokkos::complex<T1>* >
+    {
+      using common_type = typename std::common_type<T0,T1>::type;
+      using type = std::complex<common_type>;
     };
 #endif
 

